### PR TITLE
[Material] Create an InkSparkle splash effect that matches the Material 3 ripple effect

### LINKED ISSUE #82850: Introduce new Android 12 style ink ripple
Android 12 introduces a [new patterned touch effect](https://www.androidpolice.com/2021/03/24/android-12-dp-2-has-a-sweet-new-textured-ripple-animation-for-taps/) which looks different to the current `InkWell` implementation.

</img>

Flutter will match this with the upcoming Material You, but will we still be able to use the old InkWell once updated? Will we be able to use the old InkWell for older platforms and the new one starting with Android 12?

### PR DESCRIPTION
**Pull request by clocksmith**
- Create an InkSparkle effect that matches the Material 3 ripple on Android 12. Similar to InkRipple, but uses the new custom shader API.
- Uses a generated SPIR-V Fragment Shader from a GLSL source.
- Uses a generated FragmentShaderManager to make uniform updates and compilation more legible.
- Recreated from https://github.com/flutter/flutter/pull/96598
- Major step towards closing https://github.com/flutter/flutter/issues/82850 (Fixes https://github.com/flutter/flutter/issues/82850)

### Android Samples

black
pink
green

### Tests

- There are 3 paint check tests for the InkSparkle splashFactory.
 1. ElevatedButton (bounded)
 2. InkWell (bounded)
 3. InkResponse (unbounded)

- There are 18 golden tests to verify the consistency of the sparkle.
  - These tests use a deterministic version of the splashFactory, which is identical to the default splashFactory, except that it does not use `Random` to seed the turbulence effect.
  - 3 button click positions are tested: top left, center, bottom right. 
  - 6 frames are captured for each animation.

### Note on performance

Currently, the shader needs to be compiled at run time before the first InkSparkle can be painted. This takes 3-4 milliseconds on a Pixel 6 Pro and 30-35 milliseconds on a 2014 Moto E. The Moto E runs InkRipple at about 5 fps and the InkSparkle at about 5 fps. Even in the worst case, the async shader compilation is negligible, but it could be moved to MaterialApp or Theme creation instead.

### Up next

- https://github.com/flutter/flutter/issues/82850
- https://github.com/flutter/flutter/issues/73163
- https://github.com/flutter/flutter/issues/98669
- https://github.com/flutter/flutter/issues/98670
- https://github.com/flutter/flutter/issues/98671
- https://github.com/flutter/flutter/issues/99779
- https://github.com/flutter/flutter/issues/99783

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:682]: The only concern is waiting on async disk load. Happy to put there wherever it is the most painless. The compilation is already async, but I was concerned about having to load this as a web asset.

There is also a ticket to optimize the shader further to make it even less than its current ~11kb

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 20 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (zanderso) [reply packages/flutter/lib/src/material/ink_sparkle.dart:685]: The current architecture is to transpile SPIR-V to SkSL, which is then compiled by Skia at runtime. I'm suggesting to adopt the pattern of putting the data in assets to make it easier in the future (e.g. w/ Impeller which won't include a shader compiler) to shift all of the compilation work to AOT when an app is built. So loading an asset now might be a little slower, but it'll make it easier later to shift to a faster approach later.

**Comment by skia-gold** [comment]
Gold has detected about 15 new digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 7 new digest(s) on patchset 11.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:685]: Got it, since this code is currently generated, is it OK to make an issue/TODO, and follow up moving it to an asset in order to get ready for Impeller?

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 12.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:685]: https://github.com/flutter/flutter/issues/99783

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/ink_sparkle.dart:497]: I'm not that familiar with our shader integration system, but this seems like it might be better to have in a separate file to make it easier to read the surrounding code.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/ink_sparkle.dart:45]: Might want to add a "See also" pointing to the Material 3 spec for this. Also a usage example would go nice here.

**Review by darrenaustin — APPROVED** [review]
LGTM. I'll let others comment on the shader and ink features, but on the whole this looks great. Just a couple of comments below.

We can coordinate on how to hook this up in  `ThemeData` after this lands.

**Comment by skia-gold** [comment]
Gold has detected about 14 new digest(s) on patchset 13.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:45]: Is there a material 3 spec for the ripple at material.io?

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 14.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart]: Since there are no sibling files with .g.dart, I decided to keep here, but more docs explaining where it comes from

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/ink_sparkle.dart:45]: Sorry, I just assumed it would be there. I don't see one.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 16.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (zanderso) [reply packages/flutter/lib/src/material/ink_sparkle.dart:698]: I think it's fine as a follow-up PR, though I'd also like to hear thoughts from others on the reviewers list on compiling in this data. Maybe also @goderbauer

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 18.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 19.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 24 new digest(s) on patchset 20.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:27]:
```suggestion
/// component that has a `splashFactory` parameter, such as buttons:
```

File: packages/flutter/test/material/ink_sparkle_test.dart
Comment (guidezpl) [inline packages/flutter/test/material/ink_sparkle_test.dart:149]: Is there a way to avoid using a specific duration?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:502]:
```suggestion
  final double d2 = (size.topRight(Offset.zero) - size.bottomLeft(Offset.zero)).distance;
```

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:499]:
```suggestion
  final Size size = rectCallback != null ? rectCallback().size : referenceBox.size;
```

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:399]:
```suggestion
  /// in terms of "start" and "end" instead of "left" and "right".
```

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:106]: rename to `customBorderTextDirection` and make non-required?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:92]: If `radius` is derived, when/why does one pass it in?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:90]: nit: move above clipping paragraph to help readability

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/ink_sparkle.dart:389]: nit: macro this to avoid having to sync documentation with `InkSparkle`

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

Looks great! Just minor comments on readability and docs

File: packages/flutter/test/material/ink_sparkle_test.dart
Comment (clocksmith) [reply packages/flutter/test/material/ink_sparkle_test.dart:149]: In theory pumpAndSettle should be enough, but wanted to add a safety net to avoid possible flakiness

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 24.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:92]: Good question, will look at how `create` is called from factory and see what is going on here. (This affects all subclasses of `InkFeature` such as `InkRipple` and `InkSplash`. Made a note here too: https://github.com/flutter/flutter/issues/73163

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:389]: I think it makes sense to only keep this note in InkSparkle now, and then it will possibly be in one place or by way of a macro as part of https://github.com/flutter/flutter/issues/73163

**Comment by skia-gold** [comment]
Gold has detected about 24 new digest(s) on patchset 25.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:92]: Removed this doc comment to keep consistent with InkRipple and added note to https://github.com/flutter/flutter/issues/73163#issuecomment-1062744006

**Comment by skia-gold** [comment]
Gold has detected about 24 new digest(s) on patchset 29.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/test/material/ink_sparkle_test.dart
Comment (goderbauer) [inline packages/flutter/test/material/ink_sparkle_test.dart:31]: What is this test testing? (description says it's checking for drawRect, but the test doesn't seem to do that?)

File: packages/flutter/test/material/ink_sparkle_test.dart
Comment (goderbauer) [inline packages/flutter/test/material/ink_sparkle_test.dart:111]: nit: something's off with indentation here

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:23]: What about Linux?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:35]: nit: the double "then" in this sentence reads a little awkward.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:79]: add an assert for this?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:80]: Which Material is that?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:87]: Privates shouldn't be referenced in public docs.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:90]: Something got messed up with the doc here?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:101]: nit: docs should be grammatically correct and therefore sentences should start with a capital letter. May have to restructure this sentence a little bit to not have it start with the argument name.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:102]: So far, the docs haven't really talked about what the "turbulence" and "sparkles" is that this controls...

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:264]: Typically, `debugPrint` should inly be called within an assert or guarded with `kDebugMode` / `kProfileMode`. Should we do this here as well? There's probably no point in printing this in release builds...

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:388]: nit: doc sentences should start with capital letters.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:393]: Same

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [reply packages/flutter/lib/src/material/ink_sparkle.dart:691]: It is a bit strange to compile this data in like that. An asset would probably be preferred, e.g. just like we deal with the Material Design Icon font. It should probably not be included if you set "uses-material-design" to false in your pub-spec.

Follow-up PR would be fine, can you reference the issue (https://github.com/flutter/flutter/issues/99783) in a TODO somewhere here?

File: packages/flutter/test/material/ink_sparkle_test.dart
Comment (clocksmith) [reply packages/flutter/test/material/ink_sparkle_test.dart:31]: Added : expect(material, paintsExactlyCountTimes(#drawRect, 1));

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:79]: done

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:102]: Added a para on turbulence and noise, and how turbulenceSeed affects them

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:264]: changed to assert

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:80]: MaterialInkController, copied this from InkRipple, going to update docs in https://github.com/flutter/flutter/issues/73163

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:87]: fixed this

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 32.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 34.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:96]: nit: something is off in this sentence.

Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:96]: fixed

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 36.
View them at https://flutter-gold.skia.org/cl/github/99731

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (clocksmith) [reply packages/flutter/lib/src/material/ink_sparkle.dart:693]: Done

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 37.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 38.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 39.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by skia-gold** [comment]
Gold has detected about 24 new digest(s) on patchset 40.
View them at https://flutter-gold.skia.org/cl/github/99731

**Comment by zanderso** [comment]
Sorry if I'm missing this somewhere, but since @goderbauer made comments, the usual practice would have been to wait for an lgtm from him before landing.

**Comment by goderbauer** [comment]
@zanderso We synced over chat. It's all good.

**Pull request merged**
